input-profile.ts

```diff
@@ -0,0 +1,322 @@
+    }
+
+    start() {
+        this.controllerModel = null;
+        this.toFilter = new Set(['vr-mode-active-mode-switch']);
+        this.defaultControllerComponents = this.getComponents(this.defaultController);
+        this.handedness = hands[this.handednessIndex];
+
+        if (this.engine.xr?.session != null) {
+            this.engine.xr?.session.addEventListener(
+                'inputsourceschange',
+                this.onInputSourcesChange.bind(this)
+            );
+        } else {
+            this.engine.onXRSessionStart.add(() => {
+                this.engine.xr?.session.addEventListener(
+                    'inputsourceschange',
+                    this.onInputSourcesChange.bind(this)
+                );
+            });
+        }
+    }
+
+    addToFilter(element: string) {
+        this.toFilter.add(element);
+    }
+
+    removeFromFilter(element: string) {
+        this.toFilter.delete(element);
+    }
+
+    onDeactivate() {
+        if (this.engine.xr?.session != null) {
+            this.engine.xr.session.removeEventListener(
+                'inputsourceschange',
+                this.onInputSourcesChange.bind(this)
+            );
+        } else {
+            this.engine.onXRSessionStart.add(() => {
+                this.engine.xr?.session.removeEventListener(
+                    'inputsourceschange',
+                    this.onInputSourcesChange.bind(this)
+                );
+            });
+        }
+    }
+
+    setHandTrackingControllers(controllerObject: any) {
+        const HandtrackingComponent = this.trackedHand.getComponent(HandTracking);
+        if (!HandtrackingComponent) return;
+        /**@ts-ignore**/
+        HandtrackingComponent.controllerToDeactivate = controllerObject;
+    }
+
+    getComponents(obj: any) {
+        if (obj == null) return;
+        const components: Component[] = [];
+        const stack = [obj];
+        while (stack.length > 0) {
+            const currentObj = stack.pop();
+            const comps = currentObj
+                .getComponents()
+                .filter((c: any) => !this.toFilter.has(c.type));
+            components.push(...comps);
+            const children = currentObj.children || [];
+            // Push children onto the stack in reverse order to maintain the correct order
+            for (let i = children.length - 1; i >= 0; --i) {
+                stack.push(children[i]);
+            }
+        }
+        return components;
+    }
+
+    setComponentsActive(active: boolean) {
+        const comps = this.defaultControllerComponents;
+        if (!comps) return;
+        for (let i = 0; i < comps.length; ++i) {
+            comps[i].active = active;
+        }
+    }
+    // Save data to cache
+    saveToCache(key, data) {
+        try {
+            // Convert data to JSON string before storing
+            const jsonString = JSON.stringify(data);
+
+            // Save to local storage
+            localStorage.setItem(key, jsonString);
```

Comment:
Don't store that in local storage, simply store that on the class in a static object:

```js
class InputProfile {
    static Cache: Map<string, any> = new Map();
}
```